NEP: NEP 55 updates and add @mhvk as an author

ngoldbaum opened the pull request:
This marks NEP 55 as final, adds Marten as an author (I got his permission), and updates the text to match the new arena implementation from #25890.

ngoldbaum committed:
NEP: NEP 55 updates and add @mhvk as an author

mhvk commented on doc/neps/nep-0055-string_dtype.rst:
```diff
@@ -1060,16 +1064,18 @@ a "medium"-length string and the size requires only one byte in the arena
 allocation. An arena string larger than 255 bytes would need 8 bytes in the
 arena to store the size in a ``size_t``. The only flag set indicates that this
```
This is a bit confusing, with two sentences "the only flag set"

ngoldbaum replied:
Oops, remnant of the previous version. Thank you for catching.

mhvk reviewed (commented):
Nice. Some in-line comments. Also read the whole text again and noticed a few things. Most important perhaps `np.empty(3, dtype=StringDType(na_object=np.nan)` (l.409) just gives an array with 3 empty strings, not 3 nans. I don't *think* this is because of the changes I made, but we should either adjust the examples or the code... Same for the default string (l.463)

Further things I noticed (by line number):
- l.566 refers to a `SPEC-0` that I think should be a NEP (22?)
- l.570 mentions serialization. Is there actual support for this?
- l1085 says an empty string is a heap string by construction. This was and is not true: it is an uninitialized zero length string in the arena.
- l1093: I think it is more correct to say that both missing and empty strings use indications directly in the packed string and thus do not need space in the arena or on the heap.

mhvk commented:
It would not be terribly hard to get `np.empty` to give missing, though it'd mean treating `flags == 0` special (which would mean `np.zeros` also gives an array with all missing, unless we do that differently) or filling flags with `0x80` for empty (which is a bit weird for "empty" but given that we now fill with zeros, perhaps not *too* weird).

ngoldbaum commented:
> Most important perhaps np.empty(3, dtype=StringDType(na_object=np.nan) (l.409) just gives an array with 3 empty strings, not 3 nans.

This was a remnant of an earlier version of the NEP where I had that, but we later changed the design and it looks like this line never got updated. Changed to three empty strings.

> l.566 refers to a SPEC-0 that I think should be a NEP (22?)

NEP-29 has a note that it's superseded by SPEC-0. But I guess you're getting confused that I'm talking about the minimum supported numpy version. I rephrased to make that clearer.

> l.570 mentions serialization. Is there actual support for this?

Yup, you can create npy files with stringdtype arrays using the support for saving pickles.

That said, I realized reading this over just now that I missed a remnant of an earlier draft that did proposed adding full npy and npz format support in the "propsed work" section, and moved that to clearly state it's out of scope for the NEP.

> l1085 says an empty string is a heap string by construction. This was and is not true: it is an uninitialized zero length string in the arena.

Thank you, fixed.

ngoldbaum committed:
NEP: respond to marten's comments

mhvk reviewed (commented):
Thanks for making the changesso far. Note sure if you are done yet, but I noticed that the `np.empty` at l.463 is still wrong - it should give empty strings rather than 'missing' - and that l.567 still mentions SPEC-0 -- which I think is meant to be some NEP.

mhvk commented:
> Thanks for making the changesso far. Note sure if you are done yet, but I noticed that the `np.empty` at l.463 is still wrong - it should give empty strings rather than 'missing' - 

> and that l.567 still mentions SPEC-0 -- which I think is meant to be some NEP?

Oops, sorry, I see you answered that one. I had no idea was SPEC-0 was - googling it shows it is from scientific python - maybe include that?

ngoldbaum committed:
NEP: respond to marten's subsequent comments

ngoldbaum commented:
Added a hyperlink to SPEC-0 and removed the other `np.empty` example.

mhvk reviewed (approved):
Sorry that I missed the final changes - looks all good now. Get it in?

seberg commented:
Thanks @ngoldbaum and @mhvk!

seberg commented:
(NEPs are always published based of the `main` branch, so no need for backporting)
